src/Quidjibo.DataProtection/Protectors/AesPayloadProtector.cs

```diff
@@ -36,65 +42,84 @@ public async Task<byte[]> ProtectAsync(byte[] payload, CancellationToken cancell
             using (var aes = Aes.Create())
             using (var crypto = aes.CreateDecryptor(keys.CipherKey, iv))
             using (var stream = new MemoryStream())
             using (var cryptoStream = new CryptoStream(stream, crypto, CryptoStreamMode.Write))
             {
-                await cryptoStream.WriteAsync(payload, iv.Length + mac.Length, payload.Length - iv.Length - mac.Length, cancellationToken);
+                var payloadPosition = VERSION_SIZE + MAC_SIZE + IV_SIZE;
+                await cryptoStream.WriteAsync(payload, payloadPosition, payload.Length - payloadPosition, cancellationToken);
                 cryptoStream.FlushFinalBlock();
                 plaintext = stream.ToArray();
             }
 
-            VerifyMac(payload, mac, keys.MacKey);
-
             return plaintext;
         }
 
-        private byte[] ComputeMac(byte[] macKey, byte[] buffer, int offset, int count)
+        private byte[] ComputeMac(byte[] macKey, ArraySegment<byte> version, ArraySegment<byte> iv, ArraySegment<byte> ciphertext)
         {
-            using (var hmac = new HMACSHA256(macKey))
+            using (var hmac = IncrementalHash.CreateHMAC(HashAlgorithmName.SHA256, macKey))
             {
-                return hmac.ComputeHash(buffer, offset, count);
+                hmac.AppendData(version.Array, version.Offset, version.Count);
+                hmac.AppendData(iv.Array, iv.Offset, iv.Count);
+                hmac.AppendData(ciphertext.Array, ciphertext.Offset, ciphertext.Count);
+                return hmac.GetHashAndReset();
             }
         }
 
         private void VerifyMac(byte[] payload, byte[] mac, byte[] macKey)
         {
-            var payloadMac = ComputeMac(macKey, payload, 16 + 32, payload.Length - (16 + 32));
+            if (payload.Length < VERSION_SIZE + IV_SIZE + MAC_SIZE)
+            {
+                throw new MacMismatchException("MAC mismatch");
+            }
+            var versionSegment = new ArraySegment<byte>(payload, 0, VERSION_SIZE);
+            var ivSegment = new ArraySegment<byte>(payload, VERSION_SIZE, IV_SIZE);
+            var macSegment = new ArraySegment<byte>(payload, VERSION_SIZE + IV_SIZE, MAC_SIZE);
+            var payloadSegment = new ArraySegment<byte>(
+                payload,
+                VERSION_SIZE + IV_SIZE + MAC_SIZE,
+                payload.Length - (VERSION_SIZE + IV_SIZE + MAC_SIZE));
+            var payloadMac = ComputeMac(macKey, versionSegment, ivSegment, payloadSegment);
 
             if (mac.Length != payloadMac.Length)
             {
                 throw new MacMismatchException("MAC mismatch");
             }
 
-            var mismatch = false;
+            var mismatch = 0;
             for (var i = 0; i < mac.Length; i++)
             {
-                if (mac[i] != payloadMac[i])
-                {
-                    mismatch = true;
-                }
+                mismatch |= mac[i] ^ payloadMac[i];
```

Comment:
Coincidentally I happened to blog about why I changed this a few weeks ago. https://vcsjones.com/2018/10/04/fixed-time-equals-dotnet-core/

This can also hopefully be improved in netstandard2.1.